### PR TITLE
feat: Add Astra DB Vector Store Integration

### PR DESCRIPTION
**Pull request by mfortman11**
Adds a Astra DB Vector Store

Future TODOs:
- Add `delete` functionality
- Support batching for inserts

### CODE REVIEW

File: docs/core_docs/docs/integrations/vectorstores/astradb.mdx [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+export ASTRA_DB_ENDPOINT=YOUR_ASTRA_DB_ENDPOINT_HERE
+export ASTRA_DB_COLLECTION=YOUR_ASTRA_DB_COLLECTION_HERE
+export OPENAI_API_KEY=YOUR_OPENAI_API_KEY_HERE
+```
+Where `ASTRA_DB_COLLECTION` is the desired name of your collection
+
+6. Install the Astra TS Client
+
+```bash npm2yarn
+npm install @datastax/astra-db-ts
```
Comment (bracesproul) [inline docs/core_docs/docs/integrations/vectorstores/astradb.mdx:32]:
```suggestion
npm install @datastax/astra-db-ts @langchain/community
```

File: docs/core_docs/docs/integrations/vectorstores/astradb.mdx
Comment (bracesproul) [inline docs/core_docs/docs/integrations/vectorstores/astradb.mdx:1]: If you could add a note about installing other LC packages that would be great.

See [this](https://github.com/langchain-ai/langchainjs/blob/main/docs/core_docs/docs/integrations/chat/mistral.mdx#L27) as an example

File: libs/langchain-community/src/vectorstores/astradb.ts [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+      console.debug(`Collection already exists, connecting to ${this.collectionName}`);
+    }
+    this.collection = await this.astraDBClient.collection(this.collectionName);
+    console.debug("Connected to Astra DB collection");
+  }
+
+  /**
+   * Method to save vectors to AstraDB.
+   * 
+   * @todo Add batching
```
Comment (bracesproul) [inline libs/langchain-community/src/vectorstores/astradb.ts:75]: drop this from the JSDoc, we use them to generate API refs and don't want this popping up on the astradb page

**Comment by jacoblee93** [comment]
We already have one for Cassandra which I believe is compatible: 

https://js.langchain.com/docs/integrations/vectorstores/cassandra

Would this be different?

File: libs/langchain-community/src/vectorstores/astradb.ts [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    );
+
+    const topMmrMatches = mmrIndexes.map((idx) => results[idx]);
+
+    const docs: Document[] = [];
+    topMmrMatches.forEach((match) => {
+      const {
+        $similarity: _similarity, 
+        $vector: _vector, 
+        [this.idKey]: _id,
```
Comment (bracesproul) [inline libs/langchain-community/src/vectorstores/astradb.ts:214]: Why prefix with underscore?

**Comment by bracesproul** [comment]
Hey @mfortman11 I went ahead and pushed up a PR which branches off yours to resolve my comments, you can merge it into yours
https://github.com/langchain-ai/langchainjs/pull/3755
Also, I tried running the example you added and got the following errors:
```
error: c075a741-4bd9-40b5-874c-bc6ba91d66a3-us-east-2.db.astra.datastax.com/api/json/v1/default_keyspace:
error: Data: {
  createCollection: { name: 'langchain_test', options: { vector: [Object] } }
}
error: Error: Command "createCollection" failed with the following errors: [{"message":"Server call failed, please retry!"}]
    at handleIfErrorResponse (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/client/httpClient.js:226:15)
    at HTTPClient.executeCommand (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/client/httpClient.js:206:21)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at <anonymous> (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/collections/db.js:67:20)
    at executeOperation (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/collections/utils.js:161:15)
    at AstraDB.createCollection (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/collections/client.js:73:16)
    at async AstraDBVectorStore.initialize (file:///Users/bracesproul/code/lang-chain-ai/langchainjs/libs/langchain-community/dist/vectorstores/astradb.js:62:13)
    at async AstraDBVectorStore.fromDocuments (file:///Users/bracesproul/code/lang-chain-ai/langchainjs/libs/langchain-community/dist/vectorstores/astradb.js:200:9)
    at <anonymous> (/Users/bracesproul/code/lang-chain-ai/langchainjs/examples/src/indexes/vector_stores/astra.ts:25:21)
Collection already exists, connecting to langchain_test
Connected to Astra DB collection
error: c075a741-4bd9-40b5-874c-bc6ba91d66a3-us-east-2.db.astra.datastax.com/api/json/v1/default_keyspace/langchain_test:
error: Data: {
  insertMany: { documents: [ [Object], [Object], [Object] ], options: undefined }
}
error: Error: Command "insertMany" failed with the following errors: [{"message":"Server call failed, please retry!"}]
    at handleIfErrorResponse (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/client/httpClient.js:226:15)
    at HTTPClient.executeCommand (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/client/httpClient.js:206:21)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at <anonymous> (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/collections/collection.js:62:26)
    at executeOperation (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/collections/utils.js:161:15)
    at async AstraDBVectorStore.addVectors (file:///Users/bracesproul/code/lang-chain-ai/langchainjs/libs/langchain-community/dist/vectorstores/astradb.js:87:9)
    at async AstraDBVectorStore.fromDocuments (file:///Users/bracesproul/code/lang-chain-ai/langchainjs/libs/langchain-community/dist/vectorstores/astradb.js:201:9)
    at <anonymous> (/Users/bracesproul/code/lang-chain-ai/langchainjs/examples/src/indexes/vector_stores/astra.ts:25:21)
```

Any ideas?

File: libs/langchain-community/src/vectorstores/astradb.ts (hunk 3 above)
Comment (mfortman11) [reply libs/langchain-community/src/vectorstores/astradb.ts]: This was to show that these values are unused.

**Comment by mfortman11** [comment]
> Hey @mfortman11 I went ahead and pushed up a PR which branches off yours to resolve my comments, you can merge it into yours #3755 Also, I tried running the example you added and got the following errors:
> 
> ```
> error: c075a741-4bd9-40b5-874c-bc6ba91d66a3-us-east-2.db.astra.datastax.com/api/json/v1/default_keyspace:
> error: Data: {
>   createCollection: { name: 'langchain_test', options: { vector: [Object] } }
> }
> error: Error: Command "createCollection" failed with the following errors: [{"message":"Server call failed, please retry!"}]
>     at handleIfErrorResponse (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/client/httpClient.js:226:15)
>     at HTTPClient.executeCommand (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/client/httpClient.js:206:21)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at <anonymous> (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/collections/db.js:67:20)
>     at executeOperation (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/collections/utils.js:161:15)
>     at AstraDB.createCollection (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/collections/client.js:73:16)
>     at async AstraDBVectorStore.initialize (file:///Users/bracesproul/code/lang-chain-ai/langchainjs/libs/langchain-community/dist/vectorstores/astradb.js:62:13)
>     at async AstraDBVectorStore.fromDocuments (file:///Users/bracesproul/code/lang-chain-ai/langchainjs/libs/langchain-community/dist/vectorstores/astradb.js:200:9)
>     at <anonymous> (/Users/bracesproul/code/lang-chain-ai/langchainjs/examples/src/indexes/vector_stores/astra.ts:25:21)
> Collection already exists, connecting to langchain_test
> Connected to Astra DB collection
> error: c075a741-4bd9-40b5-874c-bc6ba91d66a3-us-east-2.db.astra.datastax.com/api/json/v1/default_keyspace/langchain_test:
> error: Data: {
>   insertMany: { documents: [ [Object], [Object], [Object] ], options: undefined }
> }
> error: Error: Command "insertMany" failed with the following errors: [{"message":"Server call failed, please retry!"}]
>     at handleIfErrorResponse (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/client/httpClient.js:226:15)
>     at HTTPClient.executeCommand (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/client/httpClient.js:206:21)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at <anonymous> (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/collections/collection.js:62:26)
>     at executeOperation (/Users/bracesproul/code/lang-chain-ai/langchainjs/node_modules/@datastax/astra-db-ts/dist/collections/utils.js:161:15)
>     at async AstraDBVectorStore.addVectors (file:///Users/bracesproul/code/lang-chain-ai/langchainjs/libs/langchain-community/dist/vectorstores/astradb.js:87:9)
>     at async AstraDBVectorStore.fromDocuments (file:///Users/bracesproul/code/lang-chain-ai/langchainjs/libs/langchain-community/dist/vectorstores/astradb.js:201:9)
>     at <anonymous> (/Users/bracesproul/code/lang-chain-ai/langchainjs/examples/src/indexes/vector_stores/astra.ts:25:21)
> ```
> 
> Any ideas?

It looks like the issue is with your `ASTRA_DB_ENDPOINT` it appears to be missing the `http://`
Try `https://c075a741-4bd9-40b5-874c-bc6ba91d66a3-us-east-2.db.astra.datastax.com/api/json/v1/default_keyspace`

**Comment by mfortman11** [comment]
> We already have one for Cassandra which I believe is compatible:
> 
> https://js.langchain.com/docs/integrations/vectorstores/cassandra
> 
> Would this be different?

The cassandra version was created before our JSON API was ready. The Astra version will better match the python side and be our first priority maintenance-wise

**Comment by mfortman11** [comment]
@bracesproul TY for the feedback and fixing the concerns. I made sure the example worked and believe I found you issue as commented above

**Comment by jacoblee93** [comment]
Thank you!

**Pull request merged**